### PR TITLE
Utilities: Port to LibMain

### PR DESCRIPTION
**Pull request by LucasChollet**
More ports for #11609

### CODE REVIEW

**Comment by mjz19910** [comment]
disasm needs to be updated for the lagom build as well (
`Meta/Lagom/CMakeLists.txt` ----
```
add_executable(disasm_lagom ../../Userland/Utilities/disasm.cpp)
set_target_properties(disasm_lagom PROPERTIES OUTPUT_NAME disasm)
target_link_libraries(disasm_lagom LagomCore LagomELF LagomX86)
```
)

**Comment by mjz19910** [comment]
Please rebase to include commit for `head`

File: Userland/Utilities/hostname.cpp [hunk 1]
```diff
@@ -5,27 +5,23 @@
...
-        int rc = gethostname(buffer, sizeof(buffer));
-        if (rc < 0) {
-            perror("gethostname");
-            return 1;
-        }
-        outln("{}", buffer);
+        outln("{}", TRY(Core::System::gethostname()));
     } else {
         if (strlen(hostname) >= HOST_NAME_MAX) {
             warnln("Hostname must be less than {} characters", HOST_NAME_MAX);
```
Comment (mjz19910) [inline Userland/Utilities/hostname.cpp:27]: if Core::System::sethostname does not exist, please add it... usually i find all the places that use perror, and try to remove all of them, even if i need to add support to LibCore/System.{h,cpp}

File: Userland/Utilities/host.cpp [hunk 2]
```diff
@@ -5,25 +5,22 @@
...
 
     const char* name_or_ip = nullptr;
     Core::ArgsParser args_parser;
     args_parser.set_general_help("Convert between domain name and IPv4 address.");
     args_parser.add_positional_argument(name_or_ip, "Domain name or IPv4 address", "name");
-    args_parser.parse(argc, argv);
+    args_parser.parse(args);
 
     // If input looks like an IPv4 address, we should do a reverse lookup.
     struct sockaddr_in addr;
```
Comment (mjz19910) [inline Userland/Utilities/host.cpp:26]: same here, please add Core::System::gethostbyaddr and gethostbyname

**Review by mjz19910 — CHANGES_REQUESTED** [review]
Make sure all the apps don't have lagom support that you broke!

Please separate each app into its own commit (head is mixed into gunzip).

File: Userland/Libraries/LibCore/System.cpp [hunk 3]
```diff
@@ -336,7 +337,7 @@ ErrorOr<String> ptsname(int fd)
 
 ErrorOr<String> gethostname()
 {
-    char hostname[256];
+    char hostname[HOST_NAME_MAX];
```
Comment (mjz19910) [inline Userland/Libraries/LibCore/System.cpp:340]: This does not currently work on macOS (check macOS CI run)

**Review by mjz19910 — COMMENTED** [review]
This should be ok, the hostname max len is currently missing on macOS, should probably fix limits.h before using `HOST_NAME_MAX`
```c++
#ifndef HOST_NAME_MAX
#define HOST_NAME_MAX 256
#endif
```
If there was no bug, then i would say, this looks good... go ahead and post it for review on discord. (so it can get merged faster)

**Comment by mjz19910** [comment]
```cpp
#ifndef HOST_NAME_MAX
#ifdef __APPLE__
#define HOST_NAME_MAX 255
#else
#define HOST_NAME_MAX 64
#endif
#endif
```
From this comment, somewhat modified (https://github.com/gamelinux/passivedns/issues/89#issuecomment-287635416)

File: Userland/Utilities/disasm.cpp [hunk 4]
```diff
@@ -24,15 +25,9 @@ int main(int argc, char** argv)
...
+    args_parser.parse(args);
 
-    auto file_or_error = Core::MappedFile::map(path);
-    if (file_or_error.is_error()) {
-        warnln("Could not map file: {}", file_or_error.error());
-        return 1;
-    }
-
-    auto& file = *file_or_error.value();
+    auto& file = *TRY(Core::MappedFile::map(path));
```
Comment (IdanHo) [inline Userland/Utilities/disasm.cpp:30]: This is incorrect, you're dereferencing a temporary, which results in a use after free.

File: Userland/Utilities/gunzip.cpp [hunk 5]
```diff
@@ -41,10 +41,13 @@ ErrorOr<int> serenity_main(Main::Arguments args)
         keep_input_files = true;
```
Comment (IdanHo) [inline Userland/Utilities/gunzip.cpp:41]: `assignation` does not mean what you think it means :) the correct word here would be assignment

File: Userland/Utilities/gunzip.cpp [hunk 6]
```diff
@@ -41,10 +41,13 @@ ErrorOr<int> serenity_main(Main::Arguments args)
...
     for (auto filename : filenames) {
-        if (!filename.ends_with(".gz"))
-            filename = String::formatted("{}.gz", filename);
 
-        const auto input_filename = filename;
+        const auto input_filename = [filename]() -> String {
+            if (!filename.ends_with(".gz"))
+                return String::formatted("{}.gz", filename);
+            return filename;
+        }();
```
Comment (IdanHo) [inline Userland/Utilities/gunzip.cpp:49]: What's the point of the immediately called lambda here? wouldn't this just work:
```suggestion
        String input_filename;
        if (filename.ends_with(".gz"))
            input_filename = filename;
        else
            input_filename = String::formatted("{}.gz", filename);
```

File: Userland/Libraries/LibCore/System.cpp [hunk 7]
```diff
@@ -10,6 +10,7 @@
 #include <AK/Vector.h>
```
Comment (IdanHo) [inline Userland/Libraries/LibCore/System.cpp:10]: can we mention where in LibCore the hardcoded constant is used? a la "Remove usage of a hardcoded constant in gethostname"

**Review by IdanHo — CHANGES_REQUESTED** [review]
Looks mostly good, a couple of nitpicks and one actual issue

File: Userland/Utilities/gunzip.cpp (hunk 6 above)
Comment (LucasChollet) [reply Userland/Utilities/gunzip.cpp:49]: It works, but the lambda improve const-correctness without any computational overhead.

Comment (IdanHo) [reply Userland/Utilities/gunzip.cpp:49]: > without any computational overhead
Is that based on data, or a guess? But putting that aside, having a const variable for a string (which are immutable anyways) does not seem to outweigh the cost of making the code more complicated like that.

Comment (LucasChollet) [reply Userland/Utilities/gunzip.cpp:49]: I obviously don't have any data for this particular case.
I can't retrieve the source, but I'm quite sure I got the idea from a Jason Turner's talk at CppCon.
Immutability of String doesn't prevent an assignment, whereas the const qualifier does.
Anyway, if you find it much more complicated, I will change it.

Comment (IdanHo) [reply Userland/Utilities/gunzip.cpp]: I guess it's fine, the majority of the serenity code base does not use const variables for rvalues, usually just for references and pointers, which is why I don't see the point of overcomplicating the code just to get that const in there, but I'll take your word for it that it doesn't have any adverse effects :shrug:

**Pull request merged**